spec/index.bs

```diff
@@ -4168,6 +4168,98 @@ enum GPUStoreOp {
 };
 </script>
 
+## Rasterization state ## {#render-pass-encoder-rasterization-state}
+
+The render pass encoder has several methods which affect how draw commands are rasterized to the
+attachments used by this encoder.
+
+### <dfn method for=GPURenderPassEncoder>setViewport(x, y, width, height, minDepth, maxDepth)</dfn> ### {#GPURenderPassEncoder-setViewport}
+
+<div algorithm="GPURenderPassEncoder.setViewport">
+    <strong>this:</strong> of type {{GPURenderPassEncoder}}.
+
+    **Arguments:**
+        - `float` x - Left side of the viewport in pixels
+        - `float` y - Top side of the viewport in pixels
+        - `float` |width| of the viewport in pixels
+        - `float` |height| of the viewport in pixels
+        - `float` |minDepth| - Minimum depth value of the viewport
+        - `float` |maxDepth| - Maximum depth value of the viewport. May be less than or equal to
+            |minDepth|.
+
+    **Returns:** void
+
+    Sets the viewport used during the rasterization stage to linearly map from normalized device
+    coordinates to viewport coordinates.
+
+    <div class=validusage dfn-for=GPURenderPassEncoder.setViewport>
+        <dfn abstract-op>setViewport Valid Usage</dfn>
+
+        1. |width| must be greater than `0`.
+        1. |height| must be greater than `0`.
+        1. |minDepth| must be greater than or equal to `0.0` and less than or equal to `1.0`.
+        1. |maxDepth| must be greater than or equal to `0.0` and less than or equal to `1.0`.
+
+    </div>
+
+</div>
+
+### <dfn method for=GPURenderPassEncoder>setScissorRect(x, y, width, height)</dfn> ### {#GPURenderPassEncoder-setScissorRect}
+
+<div algorithm="GPURenderPassEncoder.setScissorRect">
+    <strong>this:</strong> of type {{GPURenderPassEncoder}}.
+
+    **Arguments:**
+        - {{GPUIntegerCoordinate}} x - Left side of the scissor rectangle in pixels
+        - {{GPUIntegerCoordinate}} y - Top side of the scissor rectangle in pixels
+        - {{GPUIntegerCoordinate}} |width| of the scissor rectangle in pixels
+        - {{GPUIntegerCoordinate}} |height| of the scissor rectangle in pixels
+
+    **Returns:** void
+
+    Sets the scissor rectangle used during the rasterization stage. After transformation into
+    viewport coordinates any fragments which fall outside the scissor rectangle will be discarded.
+
+    <div class=validusage dfn-for=GPURenderPassEncoder.setScissorRect>
+        <dfn abstract-op>setScissorRect Valid Usage</dfn>
+
+        1. |width| must be greater than `0`.
+        1. |height| must be greater than `0`.
```

Comment:
Should we say that the rectangle should be contained within the viewport?

Reply:
That sounds reasonable, but I'm not seeing it called out explicitly in any of the native API docs.

Also, I just realized that I don't know if the scissor extents are intended to be relative to the attachment or the viewport.

Reply:
I don't think it has to be contained within the viewport. Shouldn't it be valid to change the viewport without changing the scissor (so then the scissor is the size of the attachment, but the viewport is smaller)?

Reply:
ok, I mean the render targets full area, not sure why I opted to write viewport here 😅 
IIRC, Metal requires scissor to be in range of the full rect

Reply:
Changed the scissor validation to require that the bounds are within the attachment. This also prompted me to add `[[attachment_size]]` as a slot on the render pass encoder to make referring to it in an algorithm easier.